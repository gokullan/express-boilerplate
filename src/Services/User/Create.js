import bcrypt from 'bcrypt'

// Models
import User from '../../Models/User'

import Logger from '../../Utils/Logger/Logger'

const Create = ({ username, name, email, password }) =>
	new Promise((resolve, reject) => {
		bcrypt
			.hash(password, 10)
			.then(passwordHash => {
				const user = new User({
					username: username.toLowerCase(),
					name,
					email: email.toLowerCase(),
					password: passwordHash,
				})

				user.save()
					.then(() => {
						Logger.info('Create user success', {
							id: user._id,
						})

						resolve({ id: user._id })
					})
					.catch(error => {
						Logger.error('Create user to database failed', {
							id: user._id,
							error,
						})

						reject({ code: 500 })
					})
			})
			.catch(error => {
				Logger.error(
					'Create user service generate password hash failed',
					{ error }
				)

				reject({ code: 500 })
			})
	})

export default Create
