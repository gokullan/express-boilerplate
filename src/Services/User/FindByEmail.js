// Models
import User from '../../Models/User'

import Logger from '../../Utils/Logger/Logger'

const FindByEmail = ({ email }) =>
	new Promise((resolve, reject) => {
		User.findOne({ email: email.toLowerCase() })
			.exec()
			.then(user => {
				resolve({ user })
			})
			.catch(error => {
				Logger.error('Find by email failed', {
					error,
				})

				reject({ code: 500 })
			})
	})

export default FindByEmail
