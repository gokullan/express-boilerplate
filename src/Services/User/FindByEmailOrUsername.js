// Models
import User from '../../Models/User'

import Logger from '../../Utils/Logger/Logger'

const FindByEmailOrUsername = ({ email, username }) =>
	new Promise((resolve, reject) => {
		User.findOne({
			$or: [
				{ email: email.toLowerCase() },
				{ username: username.toLowerCase() },
			],
		})
			.exec()
			.then(user => {
				if (!user) return reject({ code: 404 })

				resolve({ user })
			})
			.catch(error => {
				Logger.error('Find by email or username failed user service', {
					error,
				})

				reject({ code: 500 })
			})
	})

export default FindByEmailOrUsername
