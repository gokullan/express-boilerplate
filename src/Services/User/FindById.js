// Models
import User from '../../Models/User'

import Logger from '../../Utils/Logger/Logger'

const FindByEmail = ({ id }) =>
	new Promise((resolve, reject) => {
		User.findById(id)
			.exec()
			.then(user => {
				resolve({ user })
			})
			.catch(error => {
				Logger.error('Find by id failed user service', {
					error,
				})

				reject({ code: 500 })
			})
	})

export default FindByEmail
