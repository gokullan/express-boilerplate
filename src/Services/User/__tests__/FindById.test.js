import * as mockingoose from 'mockingoose'

import mongoose from 'mongoose'

import FindById from '../FindById'

// Models
import User from '../../../Models/User'

describe('Create user service', () => {
	afterEach(() => {
		mockingoose.resetAll()
	})

	it('Should create new user', () => {
		expect.assertions(6)

		const user = {
			_id: mongoose.Types.ObjectId('62c526bb503a77b155f6eba5'),
			username: 'username',
			name: 'Name',
			email: 'email@example.com',
			password: 'password',
		}

		mockingoose(User).toReturn(query => {
			expect(query.getQuery()).toMatchSnapshot('findByIdQuery')

			if (query.getQuery()._id === user._id) return user
		}, 'findOne')

		const findByIdPromise = FindById({ id: user._id }).then(
			({ user: newUser }) => {
				expect(newUser).toMatchSnapshot('findByIdResult')

				expect(newUser.username).toBe(user.username)
				expect(newUser.name).toBe(user.name)
				expect(newUser.email).toBe(user.email)
				expect(newUser._id).toBe(user._id)
			}
		)

		return findByIdPromise
	})

	it('Should reject with 404', async () => {
		expect.assertions(2)

		const user = {
			_id: mongoose.Types.ObjectId('62c526bb503a77b155f6eba6'),
			username: 'username1',
			name: 'Name1',
			email: 'email1@example.com',
			password: 'password1',
		}

		mockingoose(User).toReturn(query => {
			expect(query.getQuery()).toMatchSnapshot('findById404Query')

			if (query.getQuery()._id === user._id) return user
		}, 'findOne')

		const mock = jest.fn()

		try {
			await FindById({
				id: mongoose.Types.ObjectId('62c526bb503a77b155f6eba5'),
			})
		} catch ({ code }) {
			mock(code)
		}

		expect(mock).toHaveBeenCalledWith(404)
	})
})
