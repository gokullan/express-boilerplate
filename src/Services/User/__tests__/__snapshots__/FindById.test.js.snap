// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create user service Should create new user: findByIdQuery 1`] = `
Object {
  "_id": "62c526bb503a77b155f6eba5",
}
`;

exports[`Create user service Should create new user: findByIdResult 1`] = `
Object {
  "_id": "62c526bb503a77b155f6eba5",
  "email": "email@example.com",
  "name": "Name",
  "password": "password",
  "username": "username",
}
`;

exports[`Create user service Should reject with 404: findById404Query 1`] = `
Object {
  "_id": "62c526bb503a77b155f6eba5",
}
`;
